clang/test/CodeGenSYCL/kernel-by-reference.cpp

```diff
@@ -0,0 +1,42 @@
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2017 -DSYCL2017 %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2020 -DSYCL2020 %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -Wno-sycl-strict -DNODIAG %s
+// RUN: %clang_cc1 -triple spir64 -fsycl -fsycl-is-device -verify -fsyntax-only -sycl-std=2020 -Wno-sycl-strict -DNODIAG %s
+
+// SYCL 1.2/2017 - kernel functions passed directly. (Also no const requirement, though mutable lambdas never supported)
+template <typename name, typename Func>
+#if defined(SYCL2020)
+// expected-warning@+2 {{Passing kernel functions by value is deprecated in SYCL 2020}}
+#endif
+__attribute__((sycl_kernel)) void sycl_2017_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+// SYCL 2020 - kernel functions are passed by reference.
+template <typename name, typename Func>
+#if defined(SYCL2017)
+// expected-warning@+2 {{Passing of kernel functions by reference is a SYCL 2020 extension}}
```

Comment:
This probably should be emitted as a note or only in pedantic mode. This API can be back-ported to SYCL-1.2.1 mode (as it's done in our implementation if I understand the patch), but it might be supported by other implementations. IIRC, DPC++ compiler sets `-Wsycl-strict` if `-sycl-std=` option is set. Am I right?

I'd like to hear @erichkeane opinion on this.

```suggestion
// expected-warning@+2 {{Passing of kernel functions by reference is a SYCL 2020 extension}}
```
"extension"? I think this is standard SYCL-2020 API.